---
title: "GTSeq_AssayInfo creation"
author: "Erin D'Agnese"
date: "2023-03-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script is to make the tab deliminated text file for the GTseq optimization pipeline. What we need is something that looks like this:
Sfo_1512_18455	CATATCCTGAGACGGAAACGAT	TGGAGAGGT[A/T][GT]TCT	TGGAGACGT[A/T][GT]TCT
SNP_identifyer    Forward primer   probe1   probe2
probe1 is 7bp on either side of the SNP where probe1 is the major allele with the actual bps assignments 
probe2 is 7bp on either side of the SNP where probe2 is the minor allele with the actual bps assignments
You can have variations in all oligos e.g. [GT] for other variable BPs

```{r loading libraries}
library(here)
library(tidyverse)
library(stringr)

```

OKay so getting the name and forward primer is easy, making the probes may be a bit trickier because the fasta files that went into batchprimer3 have all ~2.8k regions and we only kept 576 for the actual SNP panel Round 1 test. So we will have to... 
1. make and bring in a file with primerID and forward primer seq in
2. bring that fasta file in, and select only those that kept in
3. find the SNP position in the sequence and pull out the 7 bp on either side (be sure to keep the alleles info)
4. copy that seq into probe1 and probe2
5. change the IUPAC code to major allele and minor which are in the fasta for probe1 and 2 respectively
6. export a tsv with name primer probe1 and probe2

1. let's start with the primer ID and forward primer seq file, use the PrimersForOrdering.csv
```{r}
AllPrimersOrdered <- read.csv(here("PrimersForOrdering.csv"))
#so there is a lot of info in here we don't need so let's filter out only forward primers and the primer IDs

AllPrimersOrdered %>%
  filter(dir == "F") %>%
  select(SNPid, Seq, Primer_type) -> PrimerIDSeqList # this is just the SNP ID and the forward primer seq and type
```

2. Bring in the fasta Okay so now we need the fasta file 
```{r}
library(Biostrings)

AllFasta <- readDNAStringSet(here("Seal_ID_SNP_panel", "SalishSea_seals_flankSNPs.fasta"), format="fasta")
#so it is a stringset but we need to convert it into a dataframe

AllFasta %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  separate(rowname, c("gnl","db","ID","pos","len","taxaID","mol","class","alleles","build"), sep = "\\|")%>%
  separate(ID, c("SNPid", "refseq"), sep = "\\s+") %>%
  mutate_at("SNPid", str_replace, "rs","") %>%
  rename_("seq" = "x") %>%
  select(SNPid, alleles, pos, seq) -> AllFastaDfSNPs

# so now we have a dataframe with the SNP ID, the major/minor alleles, and the sequence
```

```{r}
# we need to now pull out 7 bp on either side of position 100 in each seq
# one way to try this is by separating out the seq column into all the BPs separately and then select
# we have an additional problem, for the ones that are microhaplotypes we need to make those BPs if they are in the 14 bp around the SNP variable 
# there may be a faster way but since we have to make two probes and alter the bp for the SNP and the microhaps

AllFastaDfSNPs %>%
  separate(seq, 
           into = c("before","first7","SNP","second7","after"),
           sep = c(92,99,100,107)) -> AllSNPsSeqSplit

AllSNPsSeqSplit %>%
  mutate_at("alleles", str_replace, 'alleles="',"") %>%
  mutate_at("alleles", str_replace, '"', "") %>%
  separate(alleles, c("major","minor"), sep = "/") %>%
  select(SNPid, major, minor, first7, SNP, second7) -> AllSNPsSeqSplit

```

okay so we need to sort out which ones are microhaps and of those which ones do we need to pull the microhap data from so we can make those BPs variable

```{r}
# let's combine the two so we only keep the ones for the SNP panel tested

left_join(PrimerIDSeqList,AllSNPsSeqSplit, by= "SNPid") -> SNPpanelTestinfo

# now we are going to split the microhaps and SNPs out so that we can make the probes for each appropriately

SNPpanelTestinfo %>%
  filter(Primer_type == "MicoHap flanking") -> MicroHapTestinfo

SNPpanelTestinfo %>%
  filter(Primer_type == "SNP flanking") -> SNPsonlyTestinfo

```

We'll do the SNPs first since those will be easy, and will sort out the microhaps after
```{r SNP primer and probes}
SNPsonlyTestinfo %>%
  mutate(prob2first = first7) %>%
  mutate(prob2sec = second7) %>%
  unite(probe1, c(first7,major,second7), sep = "") %>%
  unite(probe2, c(prob2first,minor,prob2sec), sep = "") -> SNPsProbesInfo
#select only the columns which will be combined with the microhap ones

SNPsProbesInfo %>%
  select(SNPid, Seq, probe1, probe2) -> SNPsAssayInfo

```

Okay, so the microhaps... we had created a csv with all the microhaps grouped by hapID, so let's pull that in and only pull out the ones we kept through batchprimer3
```{r microhaps wrangling}
MicroHapsInfo <- read.csv(here("All_Microhaplotype_BP_data.csv"))
MicroHapsList <- read.csv(here("MicroHapnameList.csv"))

MicroHapsInfo %>%
  group_by(refseq,hapID) %>%
  mutate(order= 1:n()) %>%
  select(hapID,order,refseq,POS,REF,ALT) -> MicroHapsInfo2

# we need to make a column that has the appropriate SNPid for each row of the microhaps so that we can assign the relative position of the SNP and then the other vaiable allele

# so let's somewhat redo the script that pulled out the SNP which we designed the Primers off of but make the SNPid

# let's pull out the ones that have more than 2 SNPs in the microhap and then call the median row information only to add to the list to find primers for
MicroHapsInfo2 %>%
  count(refseq,hapID) -> CountHaps

MicroHapsInfo2 %>% inner_join(CountHaps, b=c("refseq","hapID")) -> Microhaps2

Microhaps2 %>% # next let's pull out the ones that only have two SNPs in each and call the 1st BP
  filter(n == 2) %>%
  mutate(SNPid = case_when(order == 1 ~ paste(refseq,"_",POS, sep = ""))) %>%
  fill(SNPid, .direction = "down") -> MicroHapsInfo2BPs # these ones will be added to the list from SNPsubset 2 ATTENTION

Microhaps2 %>% 
  filter(n > 2) %>%
  group_by(refseq,hapID) %>%
  mutate(SNPid = case_when(row_number()==ceiling(n()/2) ~ paste(refseq,"_",POS, sep = ""))) %>%
  mutate(SNPid = SNPid[!is.na(SNPid)]) %>%
  ungroup()-> MicroHapsInfoXBPhaps
# this halves the the nrow of the group, ceiling rounds up decimals for odd values

# bind the two to keep track of which ones were microhaps for downstream analysis 
MicroHapsInfo3 <- rbind(MicroHapsInfoXBPhaps,MicroHapsInfo2BPs)

# so now we can join the microhaps with the flanking seq data based on SNPid
left_join(MicroHapsInfo3, AllSNPsSeqSplit, by="SNPid") -> MicroHapsPanelInfo


```

Well we have all the info now in one dataframe for the Microhaps but we need to change the BPs to variable based on their position in the probe.
```{r microhaps BP}
# i think first it would be good to determine the BP number in relation to the SNP
MicroHapsPanelInfo %>%
  group_by(SNPid) %>%
  mutate(SNPPOS = case_when(row_number()==ceiling(n()/2) ~ 0, #assign SNP with 0
                            row_number()<=ceiling(n()/2) ~ -1,#assign anything before SNP with -1
                            row_number()>=ceiling(n()/2) ~ 1)) %>% #assign any BP after SNP with +1
  mutate(diffPrev = POS - lag(POS)) %>% # add a column with difference between SNP and prev BP in microhap
  mutate(diffNxt = POS - lead(POS)) %>% # add a column with a difference between the SNP and the next BP
  mutate(relPOS = case_when((n == 2 & SNPPOS == 1) ~ diffPrev, # if there are one 2 BP in the microhap assign the relative position to the SNP as the diffPrev which will be difference between the second BP and the prev aka SNP 
                            (n == 3 & SNPPOS < 0) ~ diffNxt, # if there are 3, the middle one is the SNP, so can take the negative difference of diffNxt
                            (n == 3 & SNPPOS > 0) ~ diffPrev, #same as prev but the value for the one after
                            (n == 4 & order == 1) ~ diffNxt, #okay so if there are 4, the 1st BP is before the SNP
                            (n == 4 & order == 3) ~ diffPrev, # the 3rd BP is the one directly after the SNP
                            (n == 4 & order == 4) ~ (diffPrev + lag(diffPrev)))) %>% # and the 4th one needs to add the differences between 4th and 3rd and the 3rd and the SNP
  ungroup() %>%
  select(SNPid, REF, ALT, SNPPOS, relPOS, first7, SNP, second7, major, minor) -> MicroHapInfoCalc


MicroHapInfoCalc %>%
  separate(first7,
           into = c("n7","n6","n5","n4","n3","n2","n1"),
           sep = c(1,2,3,4,5,6)) %>%
  separate(second7,
           into = c("p1","p2","p3","p4","p5","p6","p7"),
           sep = c(1,2,3,4,5,6)) %>%
  mutate(Insert = paste("[",REF,"/",ALT,"]", sep = "")) %>%
  mutate(n1 = case_when(relPOS == -1 ~ Insert,
                          TRUE ~ (as.character(n1)))) %>%
  mutate(n2 = case_when(relPOS == -2 ~ Insert,
                          TRUE ~ (as.character(n2)))) %>%
  mutate(n3 = case_when(relPOS == -3 ~ Insert,
                          TRUE ~ (as.character(n3)))) %>%
  mutate(n4 = case_when(relPOS == -4 ~ Insert,
                          TRUE ~ (as.character(n4)))) %>%
  mutate(n5 = case_when(relPOS == -5 ~ Insert,
                          TRUE ~ (as.character(n5)))) %>%
  mutate(n6 = case_when(relPOS == -6 ~ Insert,
                          TRUE ~ (as.character(n6)))) %>%
  mutate(n7 = case_when(relPOS == -7 ~ Insert,
                          TRUE ~ (as.character(n7)))) %>%
  mutate(p1 = case_when(relPOS == 1 ~ Insert,
                          TRUE ~ (as.character(p1)))) %>%
  mutate(p2 = case_when(relPOS == 2 ~ Insert,
                          TRUE ~ (as.character(p2)))) %>%
  mutate(p3 = case_when(relPOS == 3 ~ Insert,
                          TRUE ~ (as.character(p3)))) %>%
  mutate(p4 = case_when(relPOS == 4 ~ Insert,
                          TRUE ~ (as.character(p4)))) %>%
  mutate(p5 = case_when(relPOS == 5 ~ Insert,
                          TRUE ~ (as.character(p5)))) %>%
  mutate(p6 = case_when(relPOS == 6 ~ Insert,
                          TRUE ~ (as.character(p6)))) %>%
  mutate(p7 = case_when(relPOS == 7 ~ Insert,
                          TRUE ~ (as.character(p7)))) %>%
  mutate(relPOS = replace_na(relPOS, 0)) %>%
  filter(relPOS <= 7, relPOS >= -7) -> MicHapSepInfo


```

Okay, so now we have a dataframe where the appropriate microhap changes are changed in the rows where they were found, but we need to change those for each row of a SNPid group then we can unite them all and slice one out to be the final probe1 and probe2

```{r}
MicHapSepInfo %>%
  group_by(SNPid) %>%
  mutate(n1 = paste(unique(n1), collapse = "&")) %>%
  mutate(n2 = paste(unique(n2), collapse = "&")) %>%
  mutate(n3 = paste(unique(n3), collapse = "&")) %>%
  mutate(n4 = paste(unique(n4), collapse = "&")) %>%
  mutate(n5 = paste(unique(n5), collapse = "&")) %>%
  mutate(n6 = paste(unique(n6), collapse = "&")) %>%
  mutate(n7 = paste(unique(n7), collapse = "&")) %>%
  mutate(p1 = paste(unique(p1), collapse = "&")) %>%
  mutate(p2 = paste(unique(p2), collapse = "&")) %>%
  mutate(p3 = paste(unique(p3), collapse = "&")) %>%
  mutate(p4 = paste(unique(p4), collapse = "&")) %>%
  mutate(p5 = paste(unique(p5), collapse = "&")) %>%
  mutate(p6 = paste(unique(p6), collapse = "&")) %>%
  mutate(p7 = paste(unique(p7), collapse = "&")) -> MicHapSepInsert

nbp_cols <- c("n7","n6","n5","n4","n3","n2","n1")
pbp_cols <- c("p1","p2","p3","p4","p5","p6","p7")
MicHapSepInsert %>%
  group_by(SNPid) %>%
  mutate_at(.vars = nbp_cols, str_replace,"&.", "") %>%
  mutate_at(.vars = pbp_cols, str_replace,".&", "") %>%
  unite(first7, c(n7,n6,n5,n4,n3,n2,n1), sep = "") %>%
  unite(second7, c(p1,p2,p3,p4,p5,p6,p7), sep = "") %>%
  select(SNPid, first7, second7, major, minor) %>%
  distinct() %>%
  mutate(prob2first = first7) %>%
  mutate(prob2sec = second7) %>%
  unite(probe1, c(first7,major,second7), sep = "") %>%
  unite(probe2, c(prob2first,minor,prob2sec), sep = "") -> MicroHapsProbes
 
MicroHapTestinfo %>%
  select(SNPid, Seq) -> MicroHapsPrimers
left_join(MicroHapsPrimers, MicroHapsProbes, by= "SNPid") -> MicroHapsAssayInfo

# alright we now have the assay info format for both snps and microhaps

```

Let's merge them and export them as a TSV, remember we need to add in the sex id snps
```{r}
rbind(MicroHapsAssayInfo, SNPsAssayInfo) -> AssayInfoDF

write_tsv(AssayInfoDF, "AssayInfo.txt", col_names = FALSE)
```

