---
title: "07-GTseqRd2Opt_Correction"
author: "Erin D'Agnese"
date: "2023-05-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(here)
library(tidyverse)
```

```{r}
ReadDist <- read.table(here("Processing_GTseq_data","GTseqOptimizationRd2","Results_ReadDist.txt"), sep = "\t")

PrimeONT <- read.table(here("Processing_GTseq_data","GTseqOptimizationRd2","Results_PrimerOnT.txt"), sep = "\t")

Corrections <- read.csv(here("Processing_GTseq_data","GTseqOptimizationRd2","Results_Correction_GTperc.csv"), sep = ",", header = FALSE)
```


```{r}
#rename the columns
ReadDist %>%
  rename(SNPid = V1) %>%
  rename(MeanOn = V2) %>%
  rename(StDevOn = V3) -> ReadDist 

PrimeONT %>%
  rename(SNPid = V1) %>%
  rename(PrimeOnT = V2) %>%
  rename(StDevPOnT = V3) -> PrimeONT

Corrections %>%
  rename(SNPid = V1) %>%
  rename(Correction1 = V2) %>%
  rename(Correction2 = V3) %>%
  rename(GTPer = V4) -> Corrections

```

```{r}
# merge them by SNPid

left_join(ReadDist,PrimeONT, by = "SNPid") -> ReadsPrimeONT
left_join(ReadsPrimeONT, Corrections, by ="SNPid") -> AllRnd2Results

```

Make a Dataframe and list of the ones with 0% GT, these either will get removed or are in need of corrections
```{r}
AllRnd2Results %>%
  filter(GTPer == 0) -> ZeroGT
#some have some reads and some Primer ontarget reads, some have all zeros

# are there any that have 0 primer ont that aren't in that list, shouldn't be, but let's just check
AllRnd2Results %>%
  filter(MeanOn == 0) -> ZeroPrimeON
# okay so there are 70 with zeros across the board, let's make sure 

left_join(ZeroPrimeON, ZeroGT, by = "SNPid") # only the same 70 so we are good

anti_join(ZeroGT, ZeroPrimeON) -> PrimeOnZeroGT
# there are 17 which seem to have discrepencies which are leading to 0% GT but have on-target reads all these tho have <10 reads which get thrown out in analysis, so these can probably get removed along with the zeros across the board
```
Now let's make a list of the ones remaining after removing those
```{r}
anti_join(AllRnd2Results, ZeroGT, by = "SNPid") -> NonZeroResults
# that is 415, I also want to check that maybe we remove ones that have a Percent of primer + on-target reads that are <1% which may be worth trying to correct

NonZeroResults %>%
  filter(PrimeOnT <= 1) -> PrimeONTless1
# only another 17, so let's 

NonZeroResults %>%
  filter(PrimeOnT >= 1) -> PrimeONTGreater1 # 398 SNPs left, I think we should remove the ones with lower than 5% GT?

PrimeONTGreater1 %>%
  filter(GTPer >= 5) -> Round3Selection # there are 372 SNPs
# so I think we should select these 372 for round 3 and run samples that are from different regions to hopefully find more heterozygous alleles
```


We need to pull in the map so we can make the map for pipetting
```{r}
map <- read.csv(here("GTseq_panelOptimization_AFTC", "GTseqRd1_Goodprimers.csv"))

left_join(Round3Selection, map, by = "SNPid") -> Round3Map
anti_join(map, Round3Selection, by = "SNPid") -> Round3badprimers

Round3Map %>%
  select(SNPid, Plate, well, primer_id, GTPer) -> Round3Map

write.csv(Round3Map, "Round3PrimersFortesting.csv", row.names = FALSE)
write.csv(Round3badprimers, "Round3BadPrimers.csv", row.names = FALSE)

```

